solver/llbsolver/ops/exec.go

```diff
@@ -90,20 +90,52 @@ func cloneExecOp(old *pb.ExecOp) pb.ExecOp {
 	n.Mounts = nil
 	for i := range old.Mounts {
 		m := *old.Mounts[i]
+
+		if m.CacheOpt != nil {
+			co := *m.CacheOpt
+			m.CacheOpt = &co
+		}
+
 		n.Mounts = append(n.Mounts, &m)
 	}
 	return n
 }
 
 func (e *ExecOp) CacheMap(ctx context.Context, g session.Group, index int) (*solver.CacheMap, bool, error) {
 	op := cloneExecOp(e.op)
+
 	for i := range op.Meta.ExtraHosts {
 		h := op.Meta.ExtraHosts[i]
 		h.IP = ""
 		op.Meta.ExtraHosts[i] = h
 	}
+
+	checkShouldClearCacheOpts := func(m *pb.Mount) bool {
+		if m.CacheOpt == nil {
+			return false
+		}
+
+		if m.CacheOpt.ID == m.Dest {
```

Comment:
I think this should strictly check default ID+sharing pair. If we don't check sharing then as long as ID is equal to path, there will still be this issue that switching sharing mode invalidates the cache. We don't want to keep any special cases that continue to be broken, but we can keep the special case for strict pair because that one does not break anything and can only provide better backward compatibility.

Reply:
Done.